citation_parser.py

```diff
@@ -0,0 +1,130 @@
+import re
+from collections import namedtuple
+from types import SimpleNamespace
+from nltk.tokenize import sent_tokenize
+
+Citation = namedtuple('Citation', ['text', 'prev', 'start', 'end'])
+
+class CitationParser(object):
+
+    def __init__(self):
+        self._ref_regex = '<ref.+?(/>|</ref>)'
+        self._cn_regex = '{{(C|c)itation needed\|.+?}}'
```

Comment:
Why not use mwparserfromhell for this as well? `filter_templates` should help you find {{cn}}, and `filter_tags` will help you find the `<ref>` tags.

Reply:
I don't use mwparserfromhell because it couldn't tell me where {{cn}} or <ref> appear in the article (start and end locations). In the case, I don't want to remove {{cn}} or \<ref\>, I just need to know their locations.

Reply:
Oh right, I think you'd do something like this then:

```python
>>> import mwparserfromhell as mwp
>>> w = mwp.parse('This is a {{cn}} sentence with a template.')
>>> w.nodes
['This is a ', '{{cn}}', ' sentence with a template.']
>>> cn = w.filter_templates()[0] # find the template we want
>>> start = sum(len(n) for n in w.nodes[:w.index(cn)])
>>> end = start + len(cn)
```